base_layer/mmr/src/merkle_mountain_range.rs

```diff
@@ -0,0 +1,173 @@
+        if self.is_empty() {
+            return MerkleMountainRange::<D, B>::null_hash();
+        }
+        let peaks = find_peaks(self.hashes.len());
+        let hasher = D::new();
+        peaks
+            .into_iter()
+            .map(|i| self.hashes.get_or_panic(i))
+            .fold(hasher, |hasher, h| hasher.chain(h))
+            .result()
+            .to_vec()
+    }
+
+    /// Push a new element into the MMR. Computes new related peaks at
+    /// the same time if applicable.
+    pub fn push(&mut self, hash: &Hash) -> Result<usize, MerkleMountainRangeError> {
+        if self.is_empty() {
+            return self.push_hash(hash.clone());
+        }
+        let mut pos = self.len();
+        let (peak_map, height) = peak_map_height(pos);
+        if height != 0 {
+            return Err(MerkleMountainRangeError::CorruptDataStructure);
+        }
+        self.push_hash(hash.clone())?;
+        // hash with all immediately preceding peaks, as indicated by peak map
+        let mut peak = 1;
+        while (peak_map & peak) != 0 {
+            let left_sibling = pos + 1 - 2 * peak;
+            let left_hash = &self.hashes.get_or_panic(left_sibling);
+            peak *= 2;
+            pos += 1;
+            let last_hash = &self.hashes.get_or_panic(self.hashes.len() - 1);
+            let new_hash = hash_together::<D>(left_hash, last_hash);
+            self.push_hash(new_hash)?;
+        }
+        Ok(pos)
+    }
+
+    /// Walks the nodes in the MMR and revalidate all parent hashes
+    pub fn validate(&self) -> Result<(), MerkleMountainRangeError> {
+        // iterate on all parent nodes
+        for n in 0..self.len() {
+            let height = bintree_height(n);
+            if height > 0 {
+                let hash = self
+                    .get_node_hash(n)
+                    .ok_or(MerkleMountainRangeError::CorruptDataStructure)?;
+                let left_pos = n - (1 << height);
+                let right_pos = n - 1;
+                let left_child_hash = self
+                    .get_node_hash(left_pos)
+                    .ok_or(MerkleMountainRangeError::CorruptDataStructure)?;
+                let right_child_hash = self
+                    .get_node_hash(right_pos)
+                    .ok_or(MerkleMountainRangeError::CorruptDataStructure)?;
+                // hash the two child nodes together with parent_pos and compare
+                let hash_check = hash_together::<D>(left_child_hash, right_child_hash);
+                if &hash_check != hash {
+                    return Err(MerkleMountainRangeError::InvalidMerkleTree);
+                }
+            }
+        }
+        Ok(())
+    }
+
+    fn null_hash() -> Hash {
+        D::digest(b"").to_vec()
+    }
+
+    fn push_hash(&mut self, hash: Hash) -> Result<usize, MerkleMountainRangeError> {
+        self.hashes.push(hash).map_err(|e| {
+            error!(target: LOG_TARGET, "{:?}", e);
+            MerkleMountainRangeError::BackendPushError
```

Comment:
Is this intended? Doesn't seem to be Backend specific

Reply:
We're calling `push` on the backend, and then logging and returning an error if it fails .. that's the only place it can fail, on the backend, right?